core/jvm/src/main/scala/zio/internal/ZScheduler.scala

```diff
@@ -382,15 +365,19 @@ private final class ZScheduler(autoBlocking: Boolean) extends Executor {
                     if (size > 0) {
                       val runnables = worker.localQueue.pollUpTo(size - size / 2)
                       if (runnables.nonEmpty) {
-                        runnable = runnables.head
-                        if (runnables.tail.nonEmpty) {
-                          localQueue.offerAll(runnables.tail)
+                        var i    = 1
+                        val iter = runnables.chunkIterator
+                        runnable = iter.nextAt(0)
+                        while (iter.hasNextAt(i)) {
+                          val next1 = iter.nextAt(i)
+                          localQueue.offer(next1)
+                          i += 1
```

Comment:
Interesting. Why replacing the call to `RingBuffer#offerAll` with many `RingBuffer#offer` calls? 🤔

Reply:
What I was trying to avoid here was calling `.tail` altogether, since if the chunk has 127 elements or more, it'll materialize a new underlying Array (see [here](https://github.com/zio/zio/blob/53c4d2cc217ceddc1fece02baa69a066536c574a/core/shared/src/main/scala/zio/Chunk.scala#L303-L305)). Since here we're pulling up to 128 elements, it's fairly possible that it'll happen.

Having said that, you're quite right we should not be calling `offer` in a loop like this. To avoid the call to `.tail`, I'll add an overload to the `offerAll` in `RingBuffer` that accepts an `Iterator`. This way we get the best of both worlds :)